File: CS/12_LinksAndActions/Link.cs
using System;
using System.Drawing;
using System.Windows.Forms;
using Spire.Pdf;
using Spire.Pdf.Graphics;
using Spire.Pdf.Actions;
using Spire.Pdf.Annotations;

namespace Link
{
    public partial class Form1 : Form
    {
        public Form1()
        {
            InitializeComponent();
        }

        private void button1_Click(object sender, EventArgs e)
        {
            // Create a new PDF document.
            PdfDocument doc = new PdfDocument();

            // Set margins for the document.
            PdfUnitConvertor unitCvtr = new PdfUnitConvertor();
            PdfMargins margin = new PdfMargins();
            margin.Top = unitCvtr.ConvertUnits(2.54f, PdfGraphicsUnit.Centimeter, PdfGraphicsUnit.Point);
            margin.Bottom = margin.Top;
            margin.Left = unitCvtr.ConvertUnits(3.17f, PdfGraphicsUnit.Centimeter, PdfGraphicsUnit.Point);
            margin.Right = margin.Left;

            // Create a new page with specified size and margins.
            PdfPageBase page = doc.Pages.Add(PdfPageSize.A4, margin);

            float y = 100;
            float x = 10;

            // Set font and label for the first text link.
            PdfTrueTypeFont font = new PdfTrueTypeFont(new Font("Lucida Sans Unicode", 14));
            //////////////////Use the following code for netstandard dlls/////////////////////////
            /*
             PdfTrueTypeFont font = new PdfTrueTypeFont("Lucida Sans Unicode", 14, PdfFontStyle.Regular, true);
            */
            String label = "Simple Text Link: ";

            // Draw the label on the page.
            PdfStringFormat format = new PdfStringFormat();
            format.MeasureTrailingSpaces = true;
            page.Canvas.DrawString(label, font, PdfBrushes.Orange, 0, y, format);

            x = font.MeasureString(label, format).Width;

            // Set font and URL for the first text link.
            PdfTrueTypeFont font1 = new PdfTrueTypeFont(new Font("Lucida Sans Unicode", 14, FontStyle.Underline));
            //////////////////Use the following code for netstandard dlls/////////////////////////
            /*
            PdfTrueTypeFont font1 = new PdfTrueTypeFont("Lucida Sans Unicode", 14, PdfFontStyle.Underline, true);
            */
            String url1 = "http://www.e-iceblue.com";

            // Draw the URL on the page.
            page.Canvas.DrawString(url1, font1, PdfBrushes.CadetBlue, x, y);
            y = y + font1.MeasureString(url1).Height + 25;

            // Set font and label for the second web link.
            label = "Web Link: ";
            page.Canvas.DrawString(label, font, PdfBrushes.Orange, 0, y, format);
            x = font.MeasureString(label, format).Width;

            // Set text and properties for the second web link.
            String text = "E-iceblue home";
            PdfTextWebLink link2 = new PdfTextWebLink();
            link2.Text = text;
            link2.Url = url1;
            link2.Font = font1;
            link2.Brush = PdfBrushes.CadetBlue;

            // Draw the second web link on the page.
            link2.DrawTextWebLink(page.Canvas, new PointF(x, y));
            y = y + font1.MeasureString(text).Height + 30;

            // Set font and label for the URI annotation.
            label = "URI Annotation: ";
            page.Canvas.DrawString(label, font, PdfBrushes.Orange, 0, y, format);
            x = font.MeasureString(label, format).Width;
            text = "Google";
            PointF location = new PointF(x, y);
            SizeF size = font1.MeasureString(text);
            RectangleF linkBounds = new RectangleF(location, size);

            // Create a URI annotation and set its properties.
            PdfUriAnnotation link3 = new PdfUriAnnotation(linkBounds);
            link3.Border = new PdfAnnotationBorder(0);
            link3.Uri = "http://www.google.com";

            // Add the URI annotation to the page's annotations collection.
            (page as PdfNewPage).Annotations.Add(link3);

            // Draw the text for the URI annotation.
            page.Canvas.DrawString(text, font1, PdfBrushes.CadetBlue, x, y);
            y = y + size.Height + 30;

            // Set font and label for the URI annotation with an action.
            label = "URI Annotation Action: ";
            page.Canvas.DrawString(label, font, PdfBrushes.Orange, 0, y, format);
            x = font.MeasureString(label, format).Width;
            text = "JavaScript Action (Click Me)";
            location = new PointF(x - 2, y - 2);
            size = font1.MeasureString(text);
            size = new SizeF(size.Width + 5, size.Height + 5);
            linkBounds = new RectangleF(location, size);

            // Create a URI annotation with border and color.
            PdfUriAnnotation link4 = new PdfUriAnnotation(linkBounds);
            link4.Border = new PdfAnnotationBorder(0.75f);
            link4.Color = Color.CadetBlue;

            // Define JavaScript action script.
            String script = "app.alert({" +
                "cMsg: \"Hello.\"," +
                "nIcon: 3," +
                "cTitle: \"JavaScript Action\"" +
                "});";

            // Create a JavaScript action and assign it to the URI annotation.
            PdfJavaScriptAction action = new PdfJavaScriptAction(script);
            link4.Action = action;

            // Add the URI annotation with the JavaScript action to the page's annotations collection.
            (page as PdfNewPage).Annotations.Add(link4);

            // Draw the text for the URI annotation with the JavaScript action.
            page.Canvas.DrawString(text, font1, PdfBrushes.CadetBlue, x, y);
            y = y + size.Height + 30;

            label = "Need Help:  ";
            page.Canvas.DrawString(label, font, PdfBrushes.Orange, 0, y, format);
            x = font.MeasureString(label, format).Width;
            text = "Go to forum to ask questions";
            link2 = new PdfTextWebLink();
            link2.Text = text;
            link2.Url = "https://www.e-iceblue.com/forum/components-f5.html";
            link2.Font = font1;
            link2.Brush = PdfBrushes.CadetBlue;
            link2.DrawTextWebLink(page.Canvas, new PointF(x, y));
            y = y + font1.MeasureString(text).Height + 30;

            label = "Contct us:  ";
            page.Canvas.DrawString(label, font, PdfBrushes.Orange, 0, y, format);
            x = font.MeasureString(label, format).Width;
            text = "Send an email";
            link2 = new PdfTextWebLink();
            link2.Text = text;
            link2.Url = "mailto:support@e-iceblue.com";
            link2.Font = font1;
            link2.Brush = PdfBrushes.CadetBlue;
            link2.DrawTextWebLink(page.Canvas, new PointF(x, y));
            y = y + font1.MeasureString(text).Height + 30;

            //Save pdf file.
            doc.SaveToFile("Link.pdf");
            doc.Close();

            //Launch the file.
            PDFDocumentViewer("Link.pdf");
        }

        private void PDFDocumentViewer(string fileName)
        {
            try
            {
                System.Diagnostics.Process.Start(fileName);
            }
            catch { }
        }

    }
}

File: VB/12_LinksAndActions/Link.vb
Imports Spire.Pdf
Imports Spire.Pdf.Graphics
Imports Spire.Pdf.Actions
Imports Spire.Pdf.Annotations

Namespace Link
	Partial Public Class Form1
		Inherits Form
		Public Sub New()
			InitializeComponent()
		End Sub

		Private Sub button1_Click(ByVal sender As Object, ByVal e As EventArgs) Handles button1.Click
			' Create a new PdfDocument object
			Dim doc As New PdfDocument()

			' Create a PdfUnitConvertor object
			Dim unitCvtr As New PdfUnitConvertor()

			' Create a PdfMargins object
			Dim margin As New PdfMargins()

			' Set the top margin using converted units
			margin.Top = unitCvtr.ConvertUnits(2.54F, PdfGraphicsUnit.Centimeter, PdfGraphicsUnit.Point)

			' Set the bottom margin equal to the top margin
			margin.Bottom = margin.Top

			' Set the left margin using converted units
			margin.Left = unitCvtr.ConvertUnits(3.17F, PdfGraphicsUnit.Centimeter, PdfGraphicsUnit.Point)

			' Set the right margin equal to the left margin
			margin.Right = margin.Left

			' Add a new page to the document with specified size and margins
			Dim page As PdfPageBase = doc.Pages.Add(PdfPageSize.A4, margin)

			' Set initial coordinates for drawing text on the page
			Dim y As Single = 100
			Dim x As Single = 10

			' Create a PdfTrueTypeFont object with specified font name and size
			Dim font As New PdfTrueTypeFont(New Font("Lucida Sans Unicode", 14))
			' =============================================================================
			' Use the following code for netstandard dlls
			' =============================================================================
			'Dim font As New PdfTrueTypeFont("Lucida Sans Unicode", 14.0F, FontStyle.Regular, True)
			' =============================================================================

			' Define a label for simple text link
			Dim label As String = "Simple Text Link: "

			' Create a PdfStringFormat object
			Dim format As New PdfStringFormat()

			' Enable measurement of trailing spaces in the text
			format.MeasureTrailingSpaces = True

			' Draw the label on the page at specified coordinates with specified font and format
			page.Canvas.DrawString(label, font, PdfBrushes.Orange, 0, y, format)

			' Update the x-coordinate based on the width of the drawn label
			x = font.MeasureString(label, format).Width

			' Create a new PdfTrueTypeFont object for underlined text
			Dim font1 As New PdfTrueTypeFont(New Font("Lucida Sans Unicode", 14, FontStyle.Underline))
			' =============================================================================
			' Use the following code for netstandard dlls
			' =============================================================================
			'Dim font1 As New PdfTrueTypeFont("Lucida Sans Unicode", 14.0F, FontStyl.Underline, True)
			' =============================================================================

			' Define a URL for the web link
			Dim url1 As String = "http://www.e-iceblue.com"

			' Draw the URL on the page at specified coordinates with specified font
			page.Canvas.DrawString(url1, font1, PdfBrushes.CadetBlue, x, y)

			' Update the y-coordinate based on the height of the drawn text and add additional spacing
			y = y + font1.MeasureString(url1).Height + 25

			' Define a label for a web link
			label = "Web Link: "

			' Draw the label on the page at specified coordinates with specified font and format
			page.Canvas.DrawString(label, font, PdfBrushes.Orange, 0, y, format)

			' Update the x-coordinate based on the width of the drawn label
			x = font.MeasureString(label, format).Width

			' Define the text for the web link
			Dim text As String = "E-iceblue home"

			' Create a new PdfTextWebLink object
			Dim link2 As New PdfTextWebLink()

			' Set the properties of the web link
			link2.Text = text
			link2.Url = url1
			link2.Font = font1
			link2.Brush = PdfBrushes.CadetBlue

			' Draw the web link on the page at specified coordinates
			link2.DrawTextWebLink(page.Canvas, New PointF(x, y))

			' Update the y-coordinate based on the height of the drawn text and add additional spacing
			y = y + font1.MeasureString(text).Height + 30

			' Define a label for a URI annotation
			label = "URI Annotation: "

			' Draw the label on the page at specified coordinates with specified font and format
			page.Canvas.DrawString(label, font, PdfBrushes.Orange, 0, y, format)

			' Update the x-coordinate based on the width of the drawn label
			x = font.MeasureString(label, format).Width

			' Define the text for the URI annotation
			text = "Google"

			' Specify the location and size of the link bounds
			Dim location As New PointF(x, y)
			Dim size As SizeF = font1.MeasureString(text)
			Dim linkBounds As New RectangleF(location, size)

			' Create a new PdfUriAnnotation object with the link bounds
			Dim link3 As New PdfUriAnnotation(linkBounds)

			' Set the properties of the URI annotation
			link3.Border = New PdfAnnotationBorder(0)
			link3.Uri = "http://www.google.com"

			' Add the URI annotation to the page's annotations collection
			TryCast(page, PdfNewPage).Annotations.Add(link3)

			' Draw the text on the page at specified coordinates with specified font
			page.Canvas.DrawString(text, font1, PdfBrushes.CadetBlue, x, y)

			' Update the y-coordinate based on the height of the drawn text and add additional spacing
			y = y + size.Height + 30

			' Define a label for a URI annotation action
			label = "URI Annotation Action: "

			' Draw the label on the page at specified coordinates with specified font and format
			page.Canvas.DrawString(label, font, PdfBrushes.Orange, 0, y, format)

			' Update the x-coordinate based on the width of the drawn label
			x = font.MeasureString(label, format).Width

			' Define the text for the JavaScript action
			text = "JavaScript Action (Click Me)"

			' Specify the location and size of the link bounds
			location = New PointF(x - 2, y - 2)
			size = font1.MeasureString(text)
			size = New SizeF(size.Width + 5, size.Height + 5)
			linkBounds = New RectangleF(location, size)

			' Create a new PdfUriAnnotation object with the link bounds
			Dim link4 As New PdfUriAnnotation(linkBounds)

			' Set the properties of the URI annotation
			link4.Border = New PdfAnnotationBorder(0.75F)
			link4.Color = Color.CadetBlue

			' Define the JavaScript script
			Dim script As String = "app.alert({" & "    cMsg: ""Hello.""," & "    nIcon: 3," & "    cTitle: ""JavaScript Action""" & "});"

			' Create a new PdfJavaScriptAction object with the script
			Dim action As New PdfJavaScriptAction(script)

			' Set the action for the URI annotation
			link4.Action = action

			' Add the URI annotation to the page's annotations collection
			TryCast(page, PdfNewPage).Annotations.Add(link4)

			' Draw the text on the page at specified coordinates with specified font
			page.Canvas.DrawString(text, font1, PdfBrushes.CadetBlue, x, y)

			' Update the y-coordinate based on the height of the drawn text and add additional spacing
			y = y + size.Height + 30

			' Define a label for "Need Help"
			label = "Need Help:  "

			' Draw the label on the page at specified coordinates with specified font and format
			page.Canvas.DrawString(label, font, PdfBrushes.Orange, 0, y, format)

			' Update the x-coordinate based on the width of the drawn label
			x = font.MeasureString(label, format).Width

			' Define the text for the web link
			text = "Go to forum to ask questions"

			' Create a new PdfTextWebLink object
			link2 = New PdfTextWebLink()

			' Set the properties of the web link
			link2.Text = text
			link2.Url = "https://www.e-iceblue.com/forum/components-f5.html"
			link2.Font = font1
			link2.Brush = PdfBrushes.CadetBlue

			' Draw the web link on the page at specified coordinates
			link2.DrawTextWebLink(page.Canvas, New PointF(x, y))

			' Update the y-coordinate based on the height of the drawn text and add additional spacing
			y = y + font1.MeasureString(text).Height + 30

			' Define a label for "Contact us"
			label = "Contact us:  "

			' Draw the label on the page at specified coordinates with specified font and format
			page.Canvas.DrawString(label, font, PdfBrushes.Orange, 0, y, format)

			' Update the x-coordinate based on the width of the drawn label
			x = font.MeasureString(label, format).Width

			' Define the text for the email link
			text = "Send an email"

			' Create a new PdfTextWebLink object
			link2 = New PdfTextWebLink()

			' Set the properties of the text link
			link2.Text = text
			link2.Url = "mailto:support@e-iceblue.com"
			link2.Font = font1
			link2.Brush = PdfBrushes.CadetBlue

			' Draw the text link on the page at specified coordinates
			link2.DrawTextWebLink(page.Canvas, New PointF(x, y))

			' Update the y-coordinate based on the height of the drawn text and add additional spacing
			y = y + font1.MeasureString(text).Height + 30

			' Save the document to a file
			doc.SaveToFile("Link.pdf")

			' Close the document
			doc.Close()

			' Launch the file
			PDFDocumentViewer("Link.pdf")
		End Sub

		Private Sub PDFDocumentViewer(ByVal fileName As String)
			Try
				Process.Start(fileName)
			Catch
			End Try
		End Sub

	End Class
End Namespace
